Authorize painless execute as index action when an index is specified

javanna opened the pull request:
Painless execute allows users to validate their scripts. Some of the supported script contexts
support providing a sample document as well as an index to pull the mappings from.

The painless execute API requires cluster admin privileges today and while that's ok for the contexts that
don't support providing an index, it is not ideal when an index is provided. In fact users can run scripts
as part of the search API, which requires only the indices/read privilege on the indices that the users
is reading from.

This commit maps the painless execute action to an indices/read action when an index is specified, so that in
that case the same privileges as a search action will be requested to run painless execute.

Note: a different approach that does not require handling the painless execute case in the security code has been tried before and required creating a separate index level action with the index in the request path instead of in the request body (see #78825). We'd like to keep things as they are though in the painless execute request, with the index in the definition of the context setup as part of the request body.

Relates to #48856
Closes #86428

javanna committed:
Authorize painless execute as index action when an index is specified

Painless execute allows users to validate their scripts. Some of the supported script contexts
support providing a sample document as well as an index to pull the mappings from.

The painless execute API requires cluster admin privileges today and while that's ok for the contexts that
don't support providing an index, it is not ideal when an index is provided. In fact users can run scripts
as part of the search API, which requires only the indices/read privilege on the indices that the users
is reading from.

This commit maps the painless execute action to an indices/read action when an index is specified, so that in
that case the same privileges as a search action will be requested to run painless execute.

Relates to #48856

elasticsearchmachine commented:
Hi @javanna, I've created a changelog YAML for you.

javanna committed:
Update docs/changelog/85512.yaml

javanna commented:
I opened this as a draft to get some early feedback. If we agree that this is the way forward, it will need more tests, at least an integration test to verify that it all works as expected end-to-end.

javanna commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/SecurityActionMapper.java:
```diff
@@ -41,6 +44,13 @@ public static String action(String action, TransportRequest request) {
+                assert request instanceof SingleShardRequest;
+                String index = ((SingleShardRequest<?>) request).index();
+                if (index != null) {
+                    return INDICES_PERMISSION_PAINLESS_EXECUTE;
```
while it is pretty clear that we want to make this case an indices read action, and the request is already an indices request (it will not contain wildcards hence there is no need for it to implement Replaceable), I wonder if we are still ok with the action without the index being a cluster admin level action.

tvernum replied:
I think we're good like this.
If we made a no-index action an index privilege, then it would have to fail because we won't know how to authorize it (which index would we check?)
Leaving it as a cluster privilege will at least give it a chance to succeed.

tvernum reviewed (approved):
LGTM

mattkime commented:
addresses https://github.com/elastic/elasticsearch/issues/86428 - I wanted to make sure there was an associated issue.

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

javanna committed:
Merge branch 'main' into enhancement/painless_execute_index

javanna committed:
add integ test and fix bug

javanna commented:
@tvernum thanks for the review. I added an integration test, which made me realize I also needed to handle the shard level action. Fixed that, this should be ready now.

elasticsearchmachine commented:
Hi @javanna, I've updated the changelog YAML for you.

javanna committed:
Update docs/changelog/85512.yaml
